src/Planners/src/QuinticSpline.cpp

```diff
@@ -0,0 +1,787 @@
+                                               const std::size_t& coordinateIndex,
+                                               Eigen::Ref<Eigen::VectorXd> b)
+{
+    const auto& poly = polynomials;
+
+    const auto& i = knotIndex;
+    const auto& j = coordinateIndex;
+
+    Eigen::Matrix2d tempMatrix;
+    tempMatrix << -8 / std::pow(poly[i].duration, 2), 1 / poly[i].duration,
+                  14 / std::pow(poly[i].duration, 3), -2 / std::pow(poly[i].duration, 2);
+
+    Eigen::Vector2d tempVector;
+    tempVector << knots[i + 1].velocity[j], knots[i + 1].acceleration[j];
+    b -= tempMatrix * tempVector;
+}
+
+void QuinticSpline::Impl::addKnownTermPreviousKnot(const std::size_t& knotIndex,
+                                                   const std::size_t& coordinateIndex,
+                                                   Eigen::Ref<Eigen::VectorXd> b)
+{
+    const auto& poly = polynomials;
+
+    const auto& i = knotIndex;
+    const auto& j = coordinateIndex;
+
+    Eigen::Matrix2d tempMatrix;
+    tempMatrix << 8 / std::pow(poly[i - 1].duration, 2), 1 / poly[i - 1].duration,
+                 14 / std::pow(poly[i - 1].duration, 3), 2 / std::pow(poly[i - 1].duration, 2);
+
+    Eigen::Vector2d tempVector;
+    tempVector << knots[i - 1].velocity[j], knots[i - 1].acceleration[j];
+    b -= tempMatrix * tempVector;
+}
+
+void QuinticSpline::Impl::setPolynomialCoefficients(Polynomial& poly)
+{
+    const auto& T = poly.duration;
+
+    const auto& x0 = poly.initialPoint->position;
+    const auto& dx0 = poly.initialPoint->velocity;
+    const auto& ddx0 = poly.initialPoint->acceleration;
+
+    const auto& xT = poly.finalPoint->position;
+    const auto& dxT = poly.finalPoint->velocity;
+    const auto& ddxT = poly.finalPoint->acceleration;
+
+    poly.a0 = x0;
+    poly.a1 = dx0;
+    poly.a2 = ddx0 / 2.0;
+    poly.a3 = 1.0 / (T * T * T)
+              * (x0 * 2.0E+1 - xT * 2.0E+1 + T * dx0 * 1.2E+1 + T * dxT * 8.0 + (T * T) * ddx0 * 3.0
+                 - (T * T) * ddxT)
+              * (-1.0 / 2.0);
+    poly.a4 = (1.0 / (T * T * T * T)
+               * (x0 * 3.0E+1 - xT * 3.0E+1 + T * dx0 * 1.6E+1 + T * dxT * 1.4E+1
+                  + (T * T) * ddx0 * 3.0 - (T * T) * ddxT * 2.0))
+              / 2.0;
+    poly.a5 = 1.0 / (T * T * T * T * T)
+              * (x0 * 1.2E+1 - xT * 1.2E+1 + T * dx0 * 6.0 + T * dxT * 6.0 + (T * T) * ddx0
+                 - (T * T) * ddxT)
+              * (-1.0 / 2.0);
+}
+
+void QuinticSpline::Impl::computeIntermediateVelocitiesAndAccelaration()
+{
+    // here we assume that at least 3 points has been defined
+    const std::size_t numberOfInteriorKnots = knots.size() - 2;
+
+    Eigen::SparseMatrix<double> A(2 * numberOfInteriorKnots, 2 * numberOfInteriorKnots);
+    std::vector<Eigen::Triplet<double>> tripletsList;
+
+    std::size_t numberOfExpectedTriplets = 4;
+    if(numberOfInteriorKnots  > 1)
+    {
+        numberOfExpectedTriplets = 2 * (6 * (numberOfInteriorKnots - 2) + 2 * 4);
```

Comment:
Could you add a line on how this comes up?

Reply:
Well you are actually right the computation is a bit cryptic

given a generic internal point, the submatrix A is given by 
![image](https://user-images.githubusercontent.com/16744101/91708804-20fdc800-eb82-11ea-9174-127df3c2d33b.png) 
this submatrix has 12 entries different from zero (i.e. 12 triplets)

for the first and the last points, the submatrix A contains only 8 entries. Indeed, in the case of the first internal point, the submatrix A does not contain the first two columns, and vice-versa in the case of the last internal point,  the submatrix A does not contain the last two columns.

On the light of this the `numberOfExpectedTriplets` is equal to
`numberOfExpectedTriplets = 12 * (numberOfInteriorKnots - 2) + 8 * 2`
I can update the code in this way

```suggestion
        constexpr int numberOfNonZeroExtremalPoints = 8;
        constexpr int numberOfNonZeroInternalPoints = 12;
        constexpr int extremalPoints = 2;
        numberOfExpectedTriplets = numberOfNonZeroInternalPoints * (numberOfInteriorKnots - extremalPoints)  + numberOfNonZeroExtremalPoints * extremalPoints;
```

Reply:
I think we can simply name them `numberOfLHSExtremumElements` and `numberOfLHSNonExtremumElements` which are 4 and 6 respectively, which gives,

```cpp
numberOfExpectedTriplets = 2*(numberOfLHSNonExtremumElements*(numberOfInteriorKnots - numberOfExtremum) + 2 * numberOfLHSExtremumElements )
```

What do you think?

Reply:
I don't know if `LHS` is a good acronym in this case. Is it related to the left-hand side of the aforementioned equation?

Reply:
> Is it related to the left-hand side of the aforementioned equation?

Yes.

Reply:
> > Is it related to the left-hand side of the aforementioned equation?
> 
> Yes.

Ok because in that part of the code it is not easy to understand what is the right-hand side of the equation

Reply:
Then I think just the distinction between extremum and nonExtremum elements would be sufficient.

Reply:
Because mentioning them as,
```
constexpr int numberOfNonZeroExtremalPoints = 8;
constexpr int numberOfNonZeroInternalPoints = 12;
```

is a bit confusing as well.

Reply:
```cpp
constexpr int numberOfExtremumElement = 8;
constexpr int numberOfNonExtremumElement = 12;
```

what do you think?

Reply:
What I meant was, 
since the number of columns is 4 or 6 depending on if its an extremum or a non-extremum, we say 
```
constexpr int numberOfExtremumElements = 4;
constexpr int numberOfNonExtremumElements= 6;
```

and then we compute the `numberOfExpectedTriplets` corresponding to `x` and `xdot` as,

```
numberOfExpectedTriplets = 2*(numberOfNonExtremumElements*(numberOfInteriorKnots - numberOfExtremum) + 2 * numberOfExtremumElements )
```

Sorry for being picky here. Do you think this will be a more readable mapping of the expected matrix?

Reply:
Or do you think it's more meaningful to define them as in https://github.com/dic-iit/bipedal-locomotion-framework/pull/83#discussion_r482871832

Reply:
> ```
> fNonExtremumElements*(numberOfInteriorKnots - numberOfExtremum) + 2 * numberOfExtremumElements )
> ```
> 
> 
> Sorry for being picky here. Do you think this will be a more readable m

This is fine. I'm working on this

Reply:
I love the A :joy:

Reply:
Lovely documentation!!! Respect :rocket:

More love for the A :heart:

Reply:
Minor changes https://github.com/dic-iit/bipedal-locomotion-framework/commit/7d163f19c9ec8af0a3b413240f1bfa4605c66d3b#r42012518